x/warp/README.md

```diff
@@ -0,0 +1,190 @@
+
+To validate and deliver the message, an off-chain component delivers the signed message from Blockchain A to Blockchain B (this is out of scope for Avalanche Warp Messaging itself).
+
+When Blockchain B receives the message, it validates and delivers/executes the message:
+
+1. Read the SourceChainID of the signed message (Blockchain A)
+2. Look up the SubnetID that validates Blockchain A: Subnet A
+3. Look up the validator set of Subnet A and the registered BLS Public Keys of Subnet A at the P-Chain height specified by the ProposerVM header
+4. Filter the validators of Subnet A to include only the validators that the signed message claims as signers
+5. Verify the claimed included validators represent a sufficient quorum of stake to verify the message
+6. Aggregate the BLS Public Keys of the claimed signers into an aggregated BLS Public Key
+7. Validate the aggregate signature matches the claimed aggregate BLS Public Key
+
+After verifying the message, Blockchain B can define its own semantics of how to deliver or execute the message.
+
+### Subnet to C-Chain (Primary Network)
+
+Subnet to C-Chain Warp Messaging is a special case of Avalanche Warp Messaging. The C-Chain can efficiently verify a signature produced by another Subnet, so the implementation for the C-Chain to receive an Avalanche Warp Message can and will use the same code as is planned for Subnet-EVM.
+
+### C-Chain to Subnet
+
+To support C-Chain to Subnet communication, or more generally Primary Network to Subnet communication, we special case the C-Chain for two reasons:
+
+1. Every Subnet validator validates the C-Chain
+2. The Primary Network has the largest possible number of validators
+
+Since the Primary Network has the largest possible number of validators for any Subnet on Avalanche, it would also be the most expensive Subnet to verify Avalanche Warp Messages from (most signatures required to verify it). Luckily, we can do something much smarter.
+
+When a Subnet receives a message from a blockchain on the Primary Network, we use the validator set of the receiving Subnet instead of the entire network when validating the message. This means that the C-Chain sending a message can be the exact same as Subnet to Subnet communciation.
+
+However, when Subnet B receives a message from the C-Chain, it changes the semantics to the following:
+
+1. Read the SourceChainID of the signed message (C-Chain)
+2. Look up the SubnetID that validates C-Chain: Primary Network
+3. Look up the validator set of Subnet B (instead of the Primary Network) and the registered BLS Public Keys of Subnet B at the P-Chain height specified by the ProposerVM header
+4. Filter the validators of Subnet B to include only the validators that the signed message claims as signers
```

Comment:
I see. I was trying to say if it would be practical to assume "everyone" has signed it or do we guarantee that B validators will receive those messages (as priority). But I don't think this is in the scope at the moment?

Reply:
We do not require everyone has signed it only a threshold of stake.